### PR TITLE
Update iterBinarySearch and add test cases

### PR DESCRIPTION
**Pull request by Crazy3lf**
This is an update to my previous PR #245 .
I am not sure what name the test cases should be so I just use the same name for those cases.

### CODE REVIEW

**Comment by cclauss** [comment]
Please resolve the conflict.

**Comment by cclauss** [comment]
Please resolve conflicting files.

**Comment by cclauss** [comment]
@tjgurwara99 Can I please get your review of this PR?

**Comment by tjgurwara99** [comment]
@cclauss The code seems to be fine, do you think there should be a benchmark test for these algorithms?

**Review by tjgurwara99 — CHANGES_REQUESTED** [review]
The code is fine but I think there should also be a benchmark test in the test file for people to compare the efficiency of these algorithms.

File: searches/binarysearch.go
Comment (tjgurwara99) [inline searches/binarysearch.go:6]: Consider adding a returning a proper error by changing the return type to include the answer and an error, for example:

```
func BinarySearch(array []int, target int, lowIndex int, highIndex int) (int, error) {
   ...
}

```

and return something like this

```suggestion
return -1, errors.New("The value not found in the array")
```

File: searches/search_test.go
Comment (tjgurwara99) [inline searches/search_test.go:35]: If you add the error as a return type then it would be worth checking for the expected errors.

File: searches/binarysearch.go
Comment (Crazy3lf) [reply searches/binarysearch.go:6]: I think -1 is enough to imply not found. Is it not?

Comment (cclauss) [reply searches/binarysearch.go:6]: I agree with @tjgurwara99 on this one.  Delivering an error rather than implying an error is the correct idem in Go.

**Comment by Crazy3lf** [comment]
> 
> 
> The code is fine but I think there should also be a benchmark test in the test file for people to compare the efficiency of these algorithms.

I am rather inexperienced in writing benchmark or testing for efficiency of search algorithms so I am not sure I am capable for this😅.
Do you suggest we should have something like [this](https://github.com/golang/go/blob/618b66e16d5adedbeccb7e7ed35b26e439a9f86f/src/sort/search_test.go#L73) from Go [sort](https://golang.org/pkg/sort/) package?

File: searches/search_test.go
Comment (tjgurwara99) [inline searches/search_test.go:45]: it should be something along the lines of

```suggestion

func BenchmarkBinarySearch(b *testing.B) {
	var testCase []int
	for i := 0; i < b.N; i++ {
		testCase = append(testCase, i)
	} // You can take this completely outside with a reasonably large size say 100 - that way it is easier to take this and use it inside other benchmarks too for comparison of efficiency and speed.
	b.ResetTimer() // this is important because the above loop is expensive
	for i := 0; i < b.N; i++ {
		_, _ := BinarySearch(testCase, 10, 0, len(testCase) - 1)
	}
}

```

In the comment I meant this

```suggestion

func generateBenchmarkTestCase() []int {
	var testCase []int
	for i := 0; i < 100; i++ {
		testCase = append(testCase, i)
	}
	return testCase
}

func BenchmarkBinarySearch(b *testing.B) {
	testCase := generateBenchmarkTestCase()
	b.ResetTimer() // this is important because the generateBenchmarkTestCase() is expensive
	for i := 0; i < b.N; i++ {
		_, _ := BinarySearch(testCase, 10, 0, len(testCase) - 1)
	}
}

```

Make sure to run `gofmt -s` on the changed files for any formatting issue.

You should read up on the [testing](https://golang.org/pkg/testing/) documentation. It's quite good and very (very) useful.

In the go standard sort library that you linked, you should look at [this](https://github.com/golang/go/blob/618b66e16d5adedbeccb7e7ed35b26e439a9f86f/src/sort/search_test.go#L143) line, this is where they test the actually benchmark the sorting algorithms on in-built types.

Note that the efficiency testing that they are doing is a `test` not a `benchmark` thus it is testing something not benchmarking and you can't use cpu profilers on them (easily).

**Review by tjgurwara99 — COMMENTED** [review]
I think the changes are fine now. Once you have added the benchmark, I will approve the changes after testing.

File: searches/search_test.go
Comment (Crazy3lf) [reply searches/search_test.go:45]: Thanks for the guidance!

**Comment by raklaptudirm** [comment]
This pr currently has merge conflicts. Please fix them.

**Comment by raklaptudirm** [comment]
This pr currently has merge conflicts. Please fix them.

File: searches/search_test.go
Comment (tjgurwara99) [inline searches/search_test.go:31]: Could you move the test cases inside the test to make sure that there is no conflict with new searching algorithms that are created in this package. Also, do a git pull from our master branch and resolve conflicts. We have changed the structure of the repository which would create conflicts for this PR. For example, you can see that the package name has been changed to `search` instead of `searches` and same applies for the directory. Let me know when you are ready for this PR to be reviewed. If you do not interact with this PR for two weeks then I would consider this PR to be stale and close it. You will be able to reopen it in the case that you did not see this message.

**Review by tjgurwara99 — CHANGES_REQUESTED** [review]

File: searches/search_test.go
Comment (Crazy3lf) [reply searches/search_test.go]: The new structure have a single `testcases.go` file share across tests for different search functions. Since currently `Binary` and `BinaryIterative` function return error for when value not found but the other search functions did not, I have opted to ignore error on the test for now. Should I add a error return value for other searches?

File: search/binary_test.go
Comment (tjgurwara99) [inline search/binary_test.go:7]: I think you're currently ignoring the errors raised because of the other implementations. could you instead create a new test case locally inside this test function to check the expected errors as well. Going forward it would be good to cover everything implemented. If you can change the other implementations, I'd be happy to review but if not, I suggest that you create a local struct for the test cases and test it on them to make sure everything works as expected 😄

File: search/binary_test.go
Comment (tjgurwara99) [inline search/binary_test.go:16]: same as above ^

**Review by tjgurwara99 — CHANGES_REQUESTED** [review]
Consider the changes that I have requested. Overall, I like the changes you've made. Thanks for contributing 😄

**Review by siriak — DISMISSED** [review]
LGTM, can merge once @tjgurwara99 approves too :)

**Review by raklaptudirm — DISMISSED** [review]

File: searches/search_test.go
Comment (raklaptudirm) [reply searches/search_test.go]: @tjgurwara99

File: search/binary_test.go
Comment (Crazy3lf) [reply search/binary_test.go]: I think I will open a new PR to implement error and benchmark test on other search functions, is that ok?

**Review by siriak — APPROVED** [review]

**Review by tjgurwara99 — APPROVED** [review]
LGTM! Thanks for contributing!

**Pull request merged**
